### PR TITLE
Applicative validation

### PR DESCRIPTION
**Pull request by neshtea**
This pull requests adds the `active.clojure.validation` namespace.  It provides validators and validator combinators to build up data validation functions/expressions.

### CODE REVIEW

**Review by kaaninho — CHANGES_REQUESTED** [review]

File: README.md [hunk 1]
```diff
@@ -326,6 +326,167 @@ which can optimize work based on the equality of values.
...
+- The `mode` must be one of `:dev`, `:test`, and `:prod`
+- the `admin-users` must be a sequence of non-empty strings.
+
+First, we define a validator for ports which is not already included
+in the library:
+
+```clojure
+(defn validate-port
+  "Given a `candidate` value and an optional `label`, validates that
+  `candidate` is between ]0 65536[."
```
[the remark this answers is not in the document]
Comment (neshtea) [reply README.md:370]: You're right, thats way easier to understand.

File: README.md [hunk 2]
```diff
@@ -326,6 +326,167 @@ which can optimize work based on the equality of values.
...
+                (v/validate-one-of #{:dev :test :prod} mode :mode)
+                (v/sequence-of v/validate-non-empty-string admin-users :admin-users)))
+```
+
+We will go through the parts of this expression one by one.
+
+- `(v/validation make-config <validations>)` means that, given all
+  `<validations>` are `ValidationSuccess`es, call the function
+  `make-config` with the validated candidate values.  The
+  `make-config` function is curried automatically.
```
[the remark this answers is not in the document]
Comment (neshtea) [reply README.md:406]: You're right, the user should not need to know this.

File: src/active/clojure/validation.cljc [hunk 3]
```diff
@@ -0,0 +1,386 @@
+(ns active.clojure.validation
+  "This namespace provides the utilities for applicative data validation.
+  Chiefly, it provides the [[validation]] function that defines a data
```
[the remark this answers is not in the document]
Comment (neshtea) [reply src/active/clojure/validation.cljc:3]: I was thinking of it more as defining a value (a "validation") than of applying the validation, but maybe you're right. I would have a gripe with the other `validate-*` functions, which behave differently -- that is, they validate a candidate value.  `validation` is more of a `lift` that wraps a "normal" function into a validated one. How can we resolve this?

File: src/active/clojure/validation.cljc [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+          validations))
+
+(defn optional
+  "Takes a validation function `validate` and returns a validation
+  function that accepts what `validate` accepts plus `nil`."
+  [validate]
+  (fn [candidate & [label]]
+    (if (nil? candidate)
+      (make-validation-success candidate)
+      (let [v (validate candidate label)]
```
[the remark this answers is not in the document]
Comment (neshtea) [reply src/active/clojure/validation.cljc:298]: The `validate` function supplied in the arguments?  Sorry, I seem to not get the question.

File: src/active/clojure/validation.cljc [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+  [validators candidate & [label]]
+  ;; Choice is interesting because we need to have exactly one
+  ;; validator validate the candidate.
+  (let [validation-results   (mapv (fn [validate] (validate candidate label)) validators)
+        groups               (group-by validation-success? validation-results)
+        successes            (get groups true)
+        [failure & failures] (get groups false)]
+    (if (= 1 (count successes))
+      (first successes)
+      (reduce mappend-validation-failure failure failures))))
```
[the remark this answers is not in the document]
Comment (neshtea) [reply src/active/clojure/validation.cljc]: Thanks for catching that.  This should be resoved in 398ad50f00f2d116cacb2eff23a7679c2f283e0e.  If there is more than one error, the "choice" failure is added to the the base failures.

File: src/active/clojure/validation.cljc (hunk 4 above)
[the remark this answers is not in the document]
Comment (neshtea) [reply src/active/clojure/validation.cljc:314]: Alright, so I will mark this as resolved then.

File: src/active/clojure/validation.cljc [hunk 6]
```diff
@@ -0,0 +1,386 @@
...
+
+(defn fmap-result
+  "fmap over the result of a validation.  Applies `(fmap-success
+  f-success result`) if `result` is a [[ValidationSuccess]]
+  and `(fmap-failure f-failure result)` if `result` is
+  a [[ValidationFailure]]"
+  [f-success f-failure result]
+  (cond
+    (validation-failure? result) (fmap-failure f-success result)
+    (validation-success? result) (fmap-success f-failure result)))
```
[the remark this answers is not in the document]
Comment (neshtea) [reply src/active/clojure/validation.cljc]: This is resolved in 1c07e52cdb6b02f286c67c01e9525df31ee26d2f.

**Comment by neshtea** [comment]
> There are still "7 hidden conversations" left, did you see them, @neshtea ?

Thanks for pointing that out.  All requested changes were applied.  There is one open conversation left for you.

**Comment by kaaninho** [comment]
> All requested changes were applied.

What about this one? https://github.com/active-group/active-clojure/pull/43#discussion_r882660032 ;-)

**Review by kaaninho — APPROVED** [review]

File: test/active/clojure/validation_test.cljc [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+                                    ["d" [["e" [["f" []]]]]]]])))
+      (t/is (= (v/make-validation-failure
+                [(v/make-validation-error :e ::v/non-empty-string [[:neighbors 1] [[:neighbors 0] :label]])
+                 (v/make-validation-error :f ::v/non-empty-string [[:neighbors 1] [[:neighbors 0] [[:neighbors 0] :label]]])])
+               (validate-node ["a" [["b" [["c" []]]]
+                                    ["d" [[:e [[:f []]]]]]]]))))))
+
+(t/deftest validate-choice-test
+  (let [validate-string-or-int (fn [candidate & [label]]
+                                 (v/validate-choice [v/validate-string
```
[the remark this answers is not in the document]
Comment (neshtea) [reply test/active/clojure/validation_test.cljc:243]: Fixed in 452fd4896a489a9e359bbd94e71c2584be24a4e1

File: test/active/clojure/validation_test.cljc [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+  [label node-label
+   neighbors node-neighbors])
+
+(defn- validate-node
+  [[label neighbors]]
+  (v/validation make-node
+                (v/validate-non-empty-string label :label)
+                (v/sequence-of validate-node neighbors :neighbors)))
+
+(t/deftest sequence-of-test
```
[the remark this answers is not in the document]
Comment (neshtea) [reply test/active/clojure/validation_test.cljc:203]: Fixed in 452fd4896a489a9e359bbd94e71c2584be24a4e1

File: src/active/clojure/validation.cljc [hunk 9]
```diff
@@ -0,0 +1,424 @@
...
+
+(defn curry-n
+  "Curry a function `f` of arity `n`."
+  [f n]
+  (if (zero? n)
+    f
+    (fn [x]
+      (if (= n 1)
+        (f x)
+        (curry-n (partial f x) (dec n))))))
```
[the remark this answers is not in the document]
Comment (neshtea) [reply src/active/clojure/validation.cljc]: Nice catch.  I've added this in be151cb9d4ec4a288bb531bb2e1b490042e558e5

File: src/active/clojure/validation.cljc [hunk 10]
```diff
@@ -0,0 +1,424 @@
...
+                  
+                  (and (validation-failure? acc)
+                       (validation-failure? res))
+                  (make-validation-failure (concat (validation-failure-errors acc)
+                                                   (validation-failure-errors res)))
+                  (validation-failure? acc) acc
+
+                  (validation-failure? res) res)))
+            (pure-validation candidate)
+            validations)))
```
[the remark this answers is not in the document]
Comment (smoes) [reply src/active/clojure/validation.cljc:327]: It is far more readable for me, but I've become an "expert" for this code already :P 
Maybe leave out `juxt`, I think that is not very popular, and use the rest? But if you think it's too clever (which can be a valid point), leave it like it is.

[the remark this answers is not in the document]
Comment (smoes) [reply src/active/clojure/validation.cljc:327]: On the other hand, "collecting failures from multiple validations" is `validation` by definition, so maybe it isn't too clever

[the remark this answers is not in the document]
Comment (neshtea) [reply src/active/clojure/validation.cljc]: I've given it some thought and think you're right. I've added this in cb1cd3a85e57fa6d6e995f324ee1a466429cd97b

**Pull request merged**